### DISCUSSION TITLE
OnError handler

### DISCUSSION THREAD

**Discussion by nicjohnson145** [earthboundkid/requests#52] [Ideas] [answered]
returning an error if the validation fails is usually fine, but in some cases (like BadRequest) being able do some extra processing (like `log.Error`-ing the response body) would be quite handy. Currently in the returned error you lose access to the response body. Propose adding a `OnError` handler such that you could do something like

```go
err := requests.
        URL("api.my-site.com").
        OnError(func(origErr error, resp *http.Response) error {
            data, err := io.ReadAll(resp.Body)
            if err != nil {
                return fmt.Errorf("error reading body: %w", err)
            }
            log.Errorf("error during request: %v", data)
            return origErr
        })
```

I understand this is possible by removing the default validator, and adding a handler that calls the default validator and then does this logging on error, and then returns the default error once logged, similar to:

```go
err := requests.
        URL("api.my-site.com").
        AddValidator(nil).
        Handle(func(resp *http.Response) error {
            err := requests.DefaultValidator(resp)
            if err != nil {
                data, readErr := io.ReadAll(resp.Body)
                if readErr != nil {
                    return fmt.Errorf("error reading body: %w", readErr)
                }
                log.Errorf("error during request: %v", data)
                return err
            }
            <do some other processing, maybe using requests.ToJSON or something>
            return nil
        })
```

but it just seems like more boilerplate-y. 🤷 

This is sort of a more general handling of https://github.com/carlmjohnson/requests/discussions/32, and would be useful in cases where the object returned by the API varies based on the status code

I'd be happy to take a stab at a PR if this is something you'd be interested in.

**Comment by earthboundkid** [comment]
That’s a really interesting idea. I like having a more general way to handle error JSON than #32. It would let you choose whether to return nil error or not, for example.

Let’s make an experimental branch and play with it for a while.

**Reply by earthboundkid** [reply]
Should it run for all errors or just validation errors? I think just validation errors. But if it only runs for validation errors, it should probably be called IfInvalid or OnInvalid.

**Reply by earthboundkid** [reply]
Maybe it runs on all errors but a parameter says if it’s from prevalidation, validation, or handler.

**Reply by nicjohnson145** [reply]
I would agree with only running on validation errors. +1 for OnInvalid as the name

**Reply by nicjohnson145** [reply]
Or perhaps we have 3 methods:
* OnError - runs for both validation and handler errors
* OnInvalid - runs for validation errors
* OnHandlerError - runs for handler errors

**Reply by earthboundkid** [reply]
Maybe `OnError(kind ErrorKind, func(cause error, resp *http.Response) error)` and ErrorKind is an enum of all, connection, validation, handler.

**Reply by nicjohnson145** [reply]
That could work as well. The only downside is what if someone wants to handle only validation & handler errors (and let connection errors take the current path out). The enum approach would require they switch on the type, whereas the separate function approach would allow the handler itself to be simpler

edit: actually with the signature you specified, I'm not sure how handling a subset would work at all, unless you did some ~bitshifting~ bitmasking to represent multiple enum values, similar to the os file ~create~ open options

edit pt2: or potentially multiple handlers would require multiple calls. Once to register a handler for validation errors, and once to handle handler errors, passing the corresponding enum each time

**Reply by nicjohnson145** [reply]
A second consideration (regardless of number/signature of methods) is what happens if this handler function decides to eat the error and return nil? For example lets say it runs on validation, do the handlers still run? Or does the function stop and return nil, i.e if the handler is invoked the final return value is the value returned by the handler?

I would lean towards the second option, simply because if you wanted the handlers to run, you should modify your validation function to allow whatever situation resulted in a validation error

**Reply by earthboundkid** [reply]
I think there should be a slice of error handlers that run in the order that they're added. So maybe first you add a logger that logs all errors and then you add a validation checker that tries to recover the error JSON from a response. I think if one of the error handlers returns nil, no more error handlers run, but we don't e.g. go back and run the normal main handler because it's just too late to do that (the response has been consumed).

**Reply by nicjohnson145** [reply]
working through it, I thought a map might be a stronger choice. Opened https://github.com/carlmjohnson/requests/pull/53 as a first cut at it.

**Reply by earthboundkid** [reply]
I've been thinking about this, but haven't had time to work on it until today. Take a look at #54 and let me know what you think. Today's epiphany was that I could write a small adaptor to turn ResponseHandlers into ErrorHandlers.

**Reply by nicjohnson145** [reply]
Is the intention that if a user specifies multiple error handlers, that they would use `ErrorKind` to check if this particular handler should run? Seems like it might lead to some pretty boilerplate-y code if you wanted to register separate handlers for different kinds of errors, i.e you wanted one handler for validation errors and one handler for handler errors.

Definitely +1 on the adapter to turn ResponseHandlers into ErrorHandlers though

**Reply by earthboundkid** [reply]
It's a boilerplate tradeoff. Most of the time, you're only ever going to want either handle all errors or handle just validation errors, so it's redundant to ask every time and then pass it in as variable on top of that. I think maybe having a Builder.OnValidationError method that calls requests.ValidationHandler would cut the boilerplate down to the bare minimum. So the example from the PR would become

```go
err := requests.
 	URL("http://example.com/404").
 	ToString(&body).
 	OnError(func(err error, req *http.Request, res *http.Response) {
 		fmt.Printf("logging error (%s): %v\n",
 			requests.ErrorKindFrom(err), err)
	}).
 	OnValidationError(&handledValidationError, requests.ToString(&errBody))).
 	Fetch(context.Background())
```

All errors are logged, and the validation errors are saved into errBody.

**Reply by nicjohnson145** [reply]
Fair point, you're probably right. I can get behind that implementation.

**Reply by earthboundkid** [reply]
I'm using `OnValidationError` [in my emailalerts server](https://github.com/spotlightpa/email-alerts/blob/master/pkg/mailchimp/v3.go#L61) now.

```go
err := v3.rb.Clone().
		Pathf("/3.0/lists/%s/members/%s",
			v3.listID, SubscriberHash(req.EmailAddress)).
		Put().
		BodyJSON(&req).
		ToJSON(&resp).
		OnValidationError(nil, requests.ToJSON(&errResp)).
		Fetch(ctx)
```

I wonder if I should just make it `.OnValidationError(requests.ToJSON(&errResp))`. I think in most cases you can just use the error response if it has information and ignore it if it doesn't. The boolean out parameter is kind of awkward.

**Reply by earthboundkid** [reply]
One other thought, I dropped the thing where you could modify the error, but for logging the natural thing to do in a lot of cases is to modify the error and let the calling function actually log it, so maybe that should come back in `OnError`, which is a sort of powertool mode that most users will never use.

Also, to match `AddValidator`, the other can be called `OnValidatorError`.

**Reply by earthboundkid** [reply]
A problem with letting users modify the error is we're using the error to tell when to call OnValidatorError, so maybe that won't work.

**Reply by earthboundkid** [reply]
Still thinking about how to do this. Current thoughts:

For logging, you really do want to be able to modify the error returned to add context, and you may also want to translate all errors into [an app specific error domain](https://blog.carlmjohnson.net/post/2020/working-with-errors-as/). So the return an error design was good, but that screws up further error handlers in the chain. 

One answer is to drop the chaining and just have a single error handler. I don't like that because it means you have to choose between logging and special handling validation. Even just one error handler per type feels like a weird, artificial limitation.

So maybe the better answer to change the error handler signature to `OnError(f func(kind ErrorKind, cause error, req *http.Request, resp *http.Response) error))`. That's a long signature and the ErrorKind is semi-redundant, but it solves the chaining problem.

**Reply by earthboundkid** [reply]
I am pretty happy with the current design. The only question I have left is whether it makes sense to pass a pointer to the Builder to the error handler. My instinct is no because you can’t actually extract much information from a Builder, but OTOH, it does seem like an important bit of context.

Maybe the answer is to have a requests.ErrorParams struct and I can add new fields to it later if I need to.

**Reply by earthboundkid** [reply]
The current version of the PR has requests.OnErrorParams which turns out to be handy because you can add methods to it, like URL() which always produces a URL, even if the actual request was nil.

**Reply by earthboundkid** [reply]
My remaining question about the current API is whether I should expand it out to some kind of On-whatever hook that will run after good requests too. My gut feeling is no because that's what transports are for. But it's not trivial to make a logging transport that measures the time of a request, so maybe it would be good to have a post-hook.

**Reply by earthboundkid** [reply]
Shower thoughts: the error handling API still kind of bugs me because it adds a lot of surface area to the code. I was thinking “how could you tell which status code triggered the error handler” and it seemed like the answer is that ErrorJSON should just be its own validator, and then it knows the status already. So maybe that’s the answer, just rip out the new code and add a validator instead.

**Comment by earthboundkid** [comment, marked as answer]
Okay, #63 is merged. Thanks for all the feedback on this.
